### PR TITLE
Add "update workers" button to velum.

### PR DESCRIPTION
**Pull request by grahamhayes**
* Adds a button that will trigger the update orchestration
  run, when there is nodes that can be updated.
* Updates the status column to show status of updates.
* Updated the Salt Event listener to update status of minions
  marked as "failed" if another highstate succeeds

screenshot from 2017-06-13 13-00-48

### CODE REVIEW

File: app/controllers/concerns/discovery.rb [hunk 1]
```diff
@@ -1,18 +1,47 @@
...
+
+    minions = perform_request(endpoint: "/minions", method: "get")
+
+    #logger.info JSON.parse(minions.body)["return"]
+
+    ret = JSON.parse(minions.body)["return"]
+
+    for minion in ret do
+      minion.each do | key, value |
+        if value["tx_update_reboot_needed"]
```
Comment (ereslibre) [inline app/controllers/concerns/discovery.rb]: This `if-else` branch can be simplified as `tx_update_reboot_needed[key] = value["tx_update_reboot_needed"]`

File: app/controllers/concerns/discovery.rb [hunk 2]
```diff
@@ -1,18 +1,47 @@
...
+          tx_update_reboot_needed[key] = true        
+        else
+          tx_update_reboot_needed[key] = false
+        end
+        if value["tx_update_failed"]
```
Comment (ereslibre) [inline app/controllers/concerns/discovery.rb]: Same happens here, we can simplify it.

File: app/controllers/salt_controller.rb [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+
+    Velum::Salt.update_orchestration
+
+    Minion.where(minion_id: nil).find_each
+    respond_to do |format|
+    	format.html { redirect_to root_path }
+    	format.json { head :ok }
+    end
+	end
+end
```
Comment (ereslibre) [inline app/controllers/salt_controller.rb]: Indentation seems to be broken on the whole file.

File: app/controllers/salt_controller.rb [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+      minion.each do | key, value |
+        if value["tx_update_reboot_needed"]
+        	Minion.where(minion_id: key).update_all( highstate: Minion.highstates[:pending] )
+        end
+      end
+    end
+
+    Velum::Salt.update_orchestration
+
+    Minion.where(minion_id: nil).find_each
```
Comment (ereslibre) [inline app/controllers/salt_controller.rb]: What's the purpose of this?

File: app/assets/javascripts/dashboard.js [hunk 5]
```diff
@@ -97,10 +122,32 @@ MinionPoller = {
...
       '" value="' + minion.id + '" type="radio" disabled="" ' + checked + '>';
 
+    statusText = ''
+
+    if (tx_update_reboot_nodes && minion.highstate == "applied") {
+      statusHtml = '<i class="fa fa-arrow-circle-up text-info fa-2x" aria-hidden="true"></i>';
+      statusText = 'Update Available'
+    }
+
+    if (tx_update_reboot_nodes && minion.highstate == "failed") {
```
Comment (ereslibre) [inline app/assets/javascripts/dashboard.js]: Do we want to take this as `Update Available` and silently hide the fact  that the highstate failed?

File: app/controllers/concerns/discovery.rb [hunk 6]
```diff
@@ -1,18 +1,47 @@
...
 
   # Responds with either an HTML or JSON version of the available minions.
   def discovery
     @assigned_minions = Minion.assigned_role
     @unassigned_minions = Minion.unassigned_role
 
+    tx_update_reboot_needed = Hash.new
+    tx_update_failed = Hash.new
+
+    minions = perform_request(endpoint: "/minions", method: "get")
```
Comment (ereslibre) [inline app/controllers/concerns/discovery.rb]: I am not sure about calling all the time here to the `salt-api`. This will the request coming from the user slower. We could add a cron job that updates the status of the minions on the background and we would only ask for this column on the database (since we are missing a resque-similar tool for background jobs, that's why a cron would be the best solution IMHO).

However this can be improved at a later time, is not that important.

File: app/controllers/concerns/discovery.rb [hunk 7]
```diff
@@ -1,18 +1,47 @@
...
+
+    #logger.info JSON.parse(minions.body)["return"]
```
Comment (ereslibre) [inline app/controllers/concerns/discovery.rb]: We can remove this I guess.

File: app/controllers/salt_controller.rb [hunk 8]
```diff
@@ -0,0 +1,29 @@
...
+class SaltController < ApplicationController
+	include Velum::SaltApi
+ 	skip_before_action :redirect_to_setup
+
+	def update
+
+    minions = perform_request(endpoint: "/minions", method: "get")
+    ret = JSON.parse(minions.body)["return"]
+
+    for minion in ret do
```
Comment (ereslibre) [inline app/controllers/salt_controller.rb]: We usually use the `ret.each do |minion|` syntax, but affter using `Salt.minions` you won't need to do this I think.

File: app/controllers/concerns/discovery.rb (hunk 6 above)
Comment (ereslibre) [inline app/controllers/concerns/discovery.rb]: This can be removed to use `Salt.minions` that is already implemented: https://github.com/kubic-project/velum/blob/master/lib/velum/salt.rb#L24

File: app/controllers/salt_controller.rb [hunk 9]
```diff
@@ -0,0 +1,29 @@
+require "velum/salt"
+require "velum/salt_api"
+# HealthController holds methods for checking the readiness of the application.
+class SaltController < ApplicationController
+	include Velum::SaltApi
+ 	skip_before_action :redirect_to_setup
+
+	def update
+
+    minions = perform_request(endpoint: "/minions", method: "get")
```
Comment (ereslibre) [inline app/controllers/salt_controller.rb]: This can be removed using `Salt.minions` that is already implemented: https://github.com/kubic-project/velum/blob/master/lib/velum/salt.rb#L24

File: app/controllers/salt_controller.rb [hunk 10]
```diff
@@ -0,0 +1,29 @@
...
+    ret = JSON.parse(minions.body)["return"]
+
+    for minion in ret do
+      minion.each do | key, value |
+        if value["tx_update_reboot_needed"]
```
Comment (ereslibre) [inline app/controllers/salt_controller.rb]: Could we rename the arguments? `minion.each do |minion_id, grains|`?

File: app/controllers/salt_controller.rb [hunk 11]
```diff
@@ -0,0 +1,29 @@
...
+        	Minion.where(minion_id: key).update_all( highstate: Minion.highstates[:pending] )
```
Comment (ereslibre) [inline app/controllers/salt_controller.rb]: Would `Minion.where(minion_id: key).update_all(highstate: :pending)` work?

File: app/models/salt_handler/minion_highstate.rb [hunk 12]
```diff
@@ -25,10 +25,11 @@ def process_event
 
     # After applying a highstate once, it is applied again and again.
     # Salt probably tries to make sure it stays applied. We don't need to process
-    # those events. We only need to process a highstate if one is pending.
+    # those events. We only need to process a highstate if one is pending, or has 
+    # failed in the past.
     minion = Minion.find_by(
       minion_id: minion_id,
-      highstate: Minion.highstates[:pending]
+      highstate: [Minion.highstates[:pending], Minion.highstates[:failed]]
```
Comment (ereslibre) [inline app/models/salt_handler/minion_highstate.rb]: I also think we can directly use `highstate: [:pending, :failed]` here.

File: app/views/dashboard/index.html.slim [hunk 13]
```diff
@@ -7,15 +7,20 @@ h1 Cluster Status
         strong  new
         | nodes are available but have not been added to the cluster yet
 = link_to "Download kubectl config", kubectl_config_path, id: "download-kubeconfig", class: "btn btn-default btn-table pull-right", disabled: true
+= link_to update_path, id: "update-nodes", class: "btn btn-default btn-table pull-right", disabled: true, method: :post do
+    span.glyphicon.sidebar-icon.glyphicon-circle-arrow-up
+    span   Update Nodes
```
Comment (ereslibre) [inline app/views/dashboard/index.html.slim]: Indentation looks weird here.

File: app/controllers/concerns/discovery.rb (hunk 1 above)
Comment (ereslibre) [inline app/controllers/concerns/discovery.rb]: Also, renaming the arguments to `minion_id`, `grains` would be great, so it's something more intuitive.

File: app/controllers/salt_controller.rb (hunk 9 above)
Comment (grahamhayes) [inline app/controllers/salt_controller.rb]: done

File: app/controllers/salt_controller.rb (hunk 8 above)
Comment (grahamhayes) [inline app/controllers/salt_controller.rb]: Done

File: app/controllers/salt_controller.rb (hunk 11 above)
Comment (grahamhayes) [inline app/controllers/salt_controller.rb]: I am not sure - `Minion.highstates` is an `enum` - I have no idea how ruby deals with them

File: app/controllers/salt_controller.rb (hunk 4 above)
Comment (grahamhayes) [inline app/controllers/salt_controller.rb]: leftover code :)

File: app/assets/javascripts/dashboard.js (hunk 5 above)
Comment (grahamhayes) [inline app/assets/javascripts/dashboard.js]: I am not sure - I am open to suggestions on this

File: app/views/dashboard/index.html.slim (hunk 13 above)
Comment (grahamhayes) [inline app/views/dashboard/index.html.slim]: The spans are in the button link - I *think* they should be indented?

File: app/assets/javascripts/dashboard.js (hunk 5 above)
Comment (kiall) [inline app/assets/javascripts/dashboard.js]: Maybe use "Update Failed (Retryable)" or something like that?

File: app/controllers/salt_controller.rb (hunk 3 above)
Comment (kiall) [inline app/controllers/salt_controller.rb]: Mixing tabs and spaces @grahamhayes? I'm pretty sure that against the law (programmer law..)

File: app/controllers/salt_controller.rb [hunk 14]
```diff
@@ -0,0 +1,25 @@
+require "velum/salt"
+require "velum/salt_api"
+# HealthController holds methods for checking the readiness of the application.
```
Comment (kiall) [inline app/controllers/salt_controller.rb]: Comment is out of date

**Comment by grahamhayes** [comment]
updated-view

New screenshot

**Comment by grahamhayes** [comment]
Tested in conjunction with kubic-project/salt#105 (needed for a reboot to actually happen)

-  Added a grain to a node (manually by editing `/etc/salt/grains`).
- Ran `docker exec -it $(docker ps | grep salt-master | awk '{print $1}') salt '*' saltutil.refresh_grains`
- Opened Velum.
- Clicked "Update all Nodes"
- Watched node to see if it rebooted
- Checked velum to see if it showed no nodes needing an update

**Comment by mssola** [comment]
As @ereslibre pointed out on IRC, after #173 we might have unassigned minions on the DB, so we have to make sure that updates are not being updated when you press that button.

File: app/models/salt_handler/minion_highstate.rb (hunk 12 above)
Comment (ereslibre) [inline app/models/salt_handler/minion_highstate.rb]: This check can be removed, no? We only care about finding the `Minion` and setting it the last highstate result (so `Minion.find_by minion_id: minion_id` would be enough. (I know we want to move away from this, but I mean for 1.0 logic)

Comment (ereslibre) [inline app/models/salt_handler/minion_highstate.rb]: NVM, will merge https://github.com/kubic-project/velum/pull/189 after this and rebasing it... No need to do it here.

**Comment by ereslibre** [comment]
Rubocop is complaining about a too long example. Also, it looks like that this PR alters how the table is being built, not matching contents with columns. This is current Velum `master`:

screenshot from 2017-06-13 15-50-25

As opposed to how it looks like with this PR:

screenshot from 2017-06-13 15-48-40

**Comment by grahamhayes** [comment]
@ereslibre good catch, that was lost in rebase hell :) - Fixed
update

**Comment by ereslibre** [comment]
@grahamhayes Is the latest screenshot the last one? Or did you upload another? I'd say it still looks broken :)

File: app/assets/javascripts/dashboard.js [hunk 15]
```diff
@@ -156,10 +168,35 @@ MinionPoller = {
...
+          case "failed":
+            statusHtml = '<i class="fa fa-arrow-circle-up text-warning fa-2x" aria-hidden="true"></i>';
+            statusText = 'Update Failed - Retryable'
+            break;
+          case "pending":
+            statusText = 'Update in progress'
+            break;
+        }
+        break;
+      case 2:
```
Comment (vitoravelino) [inline app/assets/javascripts/dashboard.js]: If `update_status` is 2 then it's not retryable?

File: app/views/dashboard/index.html.slim [hunk 16]
```diff
@@ -56,15 +62,15 @@ h1 Cluster Status
...
+                | &nbsp;
+              th
                 | Status
               th
-                | Hostname
+                | ID
               th
-                | Role
+                | Hostname
               th.text-center
```
Comment (vitoravelino) [inline app/views/dashboard/index.html.slim]: You removed `Master` column here. Also, only master column needs `.text-center`, you can remove from this one.

File: app/assets/javascripts/dashboard.js [hunk 17]
```diff
@@ -156,10 +168,35 @@ MinionPoller = {
     masterHtml = '<input name="roles[master][]" id="roles_master_' + minion.id +
       '" value="' + minion.id + '" type="radio" disabled="" ' + checked + '>';
 
+    statusText = ''
```
Comment (vitoravelino) [inline app/assets/javascripts/dashboard.js]: Status text will be empty if no update? Isn't it a bit weird to have an empty "Status" column? I would guess since it's being used to show the text "update available", we could use for other states too.

File: app/views/dashboard/index.html.slim (hunk 16 above)
Comment (grahamhayes) [inline app/views/dashboard/index.html.slim]: In the mockups "Role" and "master" were combined

File: app/assets/javascripts/dashboard.js (hunk 17 above)
Comment (grahamhayes) [inline app/assets/javascripts/dashboard.js]: Thats what was shown in the mockups - so I went with what was laid out there.

File: app/assets/javascripts/dashboard.js (hunk 15 above)
Comment (grahamhayes) [inline app/assets/javascripts/dashboard.js]: No - it means the `transactional-update` has failed - which needs manual intervention

Comment (vitoravelino) [inline app/assets/javascripts/dashboard.js]: Oh. Does this apply to the admin node too?

File: app/views/dashboard/index.html.slim (hunk 16 above)
Comment (vitoravelino) [inline app/views/dashboard/index.html.slim]: You are right... Do we need the radio input then? The role is already there.

Comment (grahamhayes) [inline app/views/dashboard/index.html.slim]: We can take it out - but lets do that in a separate commit

File: app/assets/javascripts/dashboard.js (hunk 15 above)
Comment (grahamhayes) [inline app/assets/javascripts/dashboard.js]: Yes

File: app/views/dashboard/index.html.slim [hunk 18]
```diff
@@ -45,6 +49,8 @@ h1 Cluster Status
       = link_to kubectl_config_path, id: "download-kubeconfig", class: "btn btn-xs btn-default pull-right", disabled: true do
         i.fa.fa-download.fa-fw
         | kubectl config
+      a.btn.btn-default.btn-xs.update-admin-btn.pull-right.hidden data-toggle="modal" data-target=".update-admin-modal"
```
Comment (vitoravelino) [inline app/views/dashboard/index.html.slim]: You can remove this. Probably kept during rebases.

File: app/assets/javascripts/dashboard.js (hunk 15 above)
Comment (vitoravelino) [inline app/assets/javascripts/dashboard.js]: Ouch! Then the UI related to the admin node notification needs to be updated. Sorry. :(

Can I fix that in this branch? You can squash the commit before merging.

File: app/views/dashboard/index.html.slim (hunk 18 above)
Comment (grahamhayes) [inline app/views/dashboard/index.html.slim]: Done

File: app/assets/javascripts/dashboard.js (hunk 15 above)
Comment (vitoravelino) [inline app/assets/javascripts/dashboard.js]: Actually, let's confirm with @mssola. He was the one who did the `admin_status`.

**Comment by grahamhayes** [comment]
Travis looks random fails - only 2 of the ruby versions are broken - everything else is good to go

**Comment by mssola** [comment]
Related to my third point: 

lala

**Comment by mssola** [comment]
For my first point, I think that it's done because we prioritize failed updates and needed updates to the top right ? If so, ignore my first point 😉

File: app/assets/javascripts/dashboard.js (hunk 15 above)
Comment (mssola) [inline app/assets/javascripts/dashboard.js]: @vitoravelino this is a separate issue to be fixed in another PR I think. The admin node will have the same range of values. If an update has failed, imho we should be able to click the reset button. Moreover, I'd put a message somewhere telling the user that the last update of the admin node failed.

Comment (mssola) [inline app/assets/javascripts/dashboard.js]: That being said, the message should also instruct the administrator that the failure has to be manually handled. I think that we should get back to either @flavio or @eotchi on this, but as I said, I think this should be addressed in a separate PR.

File: app/controllers/salt_controller.rb [hunk 19]
```diff
@@ -0,0 +1,31 @@
+require "velum/salt"
+# HealthController holds methods for checking the readiness of the application.
+class SaltController < ApplicationController
+  skip_before_action :redirect_to_setup
+
+  def update
+    minions = assigned_with_status
+    minions.each do |minion|
+      if minion.update_status == Minion.statuses[:update_needed]
```
Comment (mssola) [inline app/controllers/salt_controller.rb]: As a general rule, avoid fat controller methods. In this case, I'd move this logic into an instance method of the `Minion` model. This method would be called in the `assigned_with_status` method when we loop through all the minions (we don't want to loop twice if we can merge both operations).

**Comment by grahamhayes** [comment]
@mssola 

> The order of the rows changes when a node has been updated. It's a minor thing, but it's still weird and unexpected that this happens.

That came in from another patchset - I am not sure how useful it is, but if we want to remove it we can make that change in a separate PR.

> Failed updates are not considered

They are - there are two types of failures.

1. `transactional-update` failed to apply. This requires manual intervention, or waiting for another `transactional-update` run to happen. We cannot run an update on a node if `tx_update_failed: True` is set.
2. The highstate failed to apply. This is possibly retryible - and the "Update" button will show in this case.

> Why should we have the text "Update Failed" if we already have the icon ?

I was working off of the mockups provided - they had this text, so I included it.

> Plus, the "Role" column is not aligned correctly.

Yup - my bad, will update that

File: app/controllers/salt_controller.rb (hunk 19 above)
Comment (grahamhayes) [inline app/controllers/salt_controller.rb]: So move the `Minion.find_by(minion_id: minion.minion_id).update(highstate: Minion.highstates[:pending])` line to a method on the model ? Or the whole `assigned_with_status` as well?

Comment (grahamhayes) [inline app/controllers/salt_controller.rb]: Updated in patchset

File: app/models/minion.rb [hunk 20]
```diff
@@ -101,6 +101,16 @@ def assign_role(new_role)
   end
   # rubocop:enable SkipsModelValidations
 
+  def self.mark_pending_update
```
Comment (mssola) [inline app/models/minion.rb]: Could you add documentation ? 😇

Comment (grahamhayes) [inline app/models/minion.rb]: Done

File: app/controllers/salt_controller.rb [hunk 21]
```diff
@@ -0,0 +1,15 @@
+require "velum/salt"
+# HealthController holds methods for checking the readiness of the application.
```
Comment (mssola) [inline app/controllers/salt_controller.rb]: Leftover ?

Comment (grahamhayes) [inline app/controllers/salt_controller.rb]: Yup - fixed

**Comment by grahamhayes** [comment]
e2e tests ran and passed locally.

**Comment by grahamhayes** [comment]
@ereslibre and @mssola had approved - I needed to retrigger CI, so that caused the +As to go away.

**Comment by mssola** [comment]
@grahamhayes can this branch be deleted, or are you using it for something else ?
